src/environments/cd.js

```diff
@@ -0,0 +1,262 @@
+import type {ParseNode, AnyParseNode} from "../parseNode";
+
+const cdArrow = {
+    ">": "\\\\cdrightarrow",
+    "<": "\\\\cdleftarrow",
+    "=": "\\\\cdlongequal",
+    "A": "\\uparrow",
+    "V": "\\downarrow",
+    "|": "\\Vert",
+    ".": "no arrow",
+};
+
+const newCell = () => {
+    return {type: "styling", body: [], mode: "math", style: "display"};
+};
+
+const isLabelEnd = (node: AnyParseNode, endChar: string): boolean => {
+    return ((node.type === "mathord" || node.type === "atom") &&
+        node.text === endChar);
+};
+
+export function parseCD(parser: Parser): ParseNode<"array"> {
+    // Get the array's parse nodes with \\ temporarily mapped to \cr.
+    const parseNodes = [];
+    parser.gullet.beginGroup();
+    parser.gullet.macros.set("\\cr", "\\\\\\relax");
+    parser.gullet.beginGroup();
+    while (true) {  // eslint-disable-line no-constant-condition
+        // Get the parse nodes for the next row.
+        parseNodes.push(parser.parseExpression(false, "\\\\"));
+        parser.gullet.endGroup();
+        parser.gullet.beginGroup();
+        const next = parser.fetch().text;
+        if (next === "&" || next === "\\\\") {
+            parser.consume();
+        } else if (next === "\\end") {
+            if (parseNodes[parseNodes.length - 1].length === 0) {
+                parseNodes.pop(); // final row ended in \\
+            }
+            break;
+        } else {
+            throw new ParseError("Expected \\\\ or \\cr or \\end",
+                                 parser.nextToken);
+        }
+    }
+
+    let row = [];
+    const body = [row];
+
+    // Loop thru the parse nodes. Collect them into cells and arrows.
+    for (let i = 0; i < parseNodes.length; i++) {
+        // Start a new row.
+        const nodes = parseNodes[i];
+        let cell = newCell();
+
+        for (let j = 0; j < nodes.length; j++) {
+            if (!(nodes[j].type === "textord" && nodes[j].text === "@")) {
+                cell.body.push(nodes[j]);
+            } else {
+                // Parse node j is an "@", the start of an arrow.
+                row.push(cell);
+                const arrowChar = assertSymbolNodeType(nodes[j + 1]).text;
+                const firstLabel = {type: "ordgroup", mode: "math", body: []};
+                const secondLabel = {type: "ordgroup", mode: "math", body: []};
+                if ("=|.".indexOf(arrowChar) > -1) {
+                    j += 1;
+                } else if ("<>AV".indexOf(arrowChar) > -1) {
+                    // Get the arrow labels.
+                    for (let k = j + 2; k < nodes.length; k++) {
+                        if (isLabelEnd(nodes[k], arrowChar)) {
+                            j = k;
+                            break;
+                        }
+                        firstLabel.body.push(nodes[k]);
+                    }
+                    for (let k = j + 1; k < nodes.length; k++) {
+                        if (isLabelEnd(nodes[k], arrowChar)) {
```

Comment:
What happens if `isLabelEnd` never return true?  Is that valid with the `CD` environment?

Reply:
Bad things happen. AMS LaTeX writes many unhelpful comments. I've just now figured out how to trap for this condition and write a good ParseError. Should be better now.